mlrun/serving/routers.py

```diff
@@ -244,7 +252,144 @@ class OperationTypes(str, Enum):
+                    routes[key] = step
+                executor_class = concurrent.futures.ProcessPoolExecutor
+                self._pool = executor_class(
+                    max_workers=len(self.routes),
+                    initializer=_ParallelRunInterface.init_pool,
+                    initargs=(
+                        server,
+                        routes,
+                    ),
+                )
+            elif self.executor_type == ParallelRunnerModes.thread:
+                executor_class = concurrent.futures.ThreadPoolExecutor
+                self._pool = executor_class(max_workers=len(self.routes))
+
+        return self._pool
+
+    def _shutdown_pool(self):
+        """
+        Shutdowns the pool and updated self._pool to None
+        """
+        if self._pool is not None:
+            self._pool.shutdown()
+            self._pool = None
+
+    def _parallel_run(self, event: dict):
+        """
+        Execute parallel run
+
+        :param event: event to run in parallel
+
+        :return: All the results of the runs
+        """
+        results = {}
+        if self.executor_type == ParallelRunnerModes.array:
+            results = {
+                model_name: model.run(copy.copy(event)).body
+                for model_name, model in self.routes.items()
+            }
+            return results
+        futures = []
+        executor = self._init_pool()
+        for route in self.routes.keys():
+            if self.executor_type == ParallelRunnerModes.process:
+                future = executor.submit(
+                    _ParallelRunInterface._wrap_step, route, copy.copy(event)
+                )
+            elif self.executor_type == ParallelRunnerModes.thread:
+                step = self.routes[route]
+                future = executor.submit(
+                    _ParallelRunInterface._wrap_method,
+                    route,
+                    step.run,
+                    copy.copy(event),
+                )
+
+            futures.append(future)
+
+        for future in concurrent.futures.as_completed(futures):
+            try:
+                key, result = future.result()
+                results[key] = result.body
+            except Exception as exc:
+                logger.error(traceback.format_exc())
+                print(f"child route generated an exception: {exc}")
+        self.context.logger.debug(f"Collected results from children: {results}")
+        return results
+
+    @staticmethod
+    def init_pool(server_spec, routes):
+        server = mlrun.serving.GraphServer.from_dict(server_spec)
+        server.init_states(None, None)
+        global local_routes
+        for route in routes.values():
+            route.context = server.context
+            if route._object:
+                route._object.context = server.context
+        local_routes = routes
+
+    @staticmethod
+    def _wrap_step(route, event):
+        return route, local_routes[route].run(event)
```

Comment:
is `local_routes` global?

Reply:
Yes, you can see it on [this line](https://github.com/mlrun/mlrun/pull/2798/files#diff-451ed0f9d5bbdd0b1a41457025207c04a5cbbefb0ca9b554de1ae128e7ccb136R376)